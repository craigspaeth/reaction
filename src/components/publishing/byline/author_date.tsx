import moment from "moment-timezone"
import * as React from "react"
import styled, { StyledFunction } from "styled-components"
import { pMedia } from "../../helpers"
import Fonts from "../fonts"

interface AuthorDateProps {
  authors?: any
  date?: string
  layout: string
}

export const Author: React.SFC<AuthorDateProps> = props => {
  return <StyledAuthor className="author" layout={props.layout}>By {getAuthorByline(props.authors)}</StyledAuthor>
}

export const Date: React.SFC<AuthorDateProps> = props => {
  return <Text className="date" layout={props.layout}>{getDate(props.date, props.layout)}</Text>
}

const getAuthorByline = authors => {
  if (authors && authors.length > 0) {
    if (authors.length > 1) {
      const str = authors.reduce((prev, curr, i) => {
        let delim
        const len = authors.length
        if (i === len - 1) {
          delim = " and "
        } else if (i === 0) {
          delim = ""
        } else {
          delim = ", "
        }
        return prev + delim + curr.name
      }, "")
      return str
    } else {
      return authors[0].name
    }
  } else {
    return "Artsy Editorial"
  }
}

const getDate = (date, layout) => {
  return layout === "condensed"
    ? moment(date).tz("America/New_York").format("MMM D, YYYY")
    : moment(date).tz("America/New_York").format("MMM D, YYYY h:mm a")
}

const div: StyledFunction<AuthorDateProps & React.HTMLProps<HTMLInputElement>> = styled.div

const Text = div`
  ${props => (props.layout === "condensed" ? Fonts.unica("s14", "medium") : Fonts.unica("s16", "medium"))}
  margin: 10px 30px 0 0;
  &.date {
    white-space: nowrap;
  }
  ${props => pMedia.sm`
    ${props.layout === "condensed" ? Fonts.unica("s12", "medium") : Fonts.unica("s14", "medium")}
    ${Fonts.unica("s14", "medium")}
    margin: 10px 20px 0 0;
  `}
`
const adjustForCondensed = layout => {
  return layout === "condensed"
    ? `
  &:before {
    min-width: 8px;
    min-height: 8px;
    margin-right: 5px;
  }`
    : ""
}
const StyledAuthor = Text.extend`
  &:before {
    content: "";
    display: inline-block;
    min-width: 10px;
    min-height: 10px;
    border-radius: 50%;
    margin: 6px 10px 0 0;
    background-color: ${props => (props.layout === "fullscreen" ? "#fff" : "#000")};
  }
  ${props => adjustForCondensed(props.layout)}
  ${pMedia.sm`
    &:before {
      min-width: 8px;
      min-height: 8px;
    }
  `}
`
