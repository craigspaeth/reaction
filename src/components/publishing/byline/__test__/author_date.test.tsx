import { shallow } from "enzyme"
import "jest-styled-components"
import * as React from "react"
import * as renderer from "react-test-renderer"
import { Author, Date } from "../author_date"

describe("Author/Date", () => {
  it("renders a single author", () => {
    const authors = [{ name: "Molly Gottschalk" }]
    const author = renderer.create(<Author authors={authors} layout={"split"} />)
    expect(author).toMatchSnapshot()
    const shallowAuthor = shallow(<Author authors={authors} layout={"split"} />)
    expect(shallowAuthor.html()).toContain("Molly Gottschalk")
  })

  it("renders multiple authors", () => {
    const authors = [{ name: "Molly Gottschalk" }, { name: "Kana Abe" }]
    const author = renderer.create(<Author authors={authors} layout={"split"} />)
    expect(author).toMatchSnapshot()
    const shallowAuthor = shallow(<Author authors={authors} layout={"split"} />)
    expect(shallowAuthor.html()).toContain("Molly Gottschalk")
    expect(shallowAuthor.html()).toContain("Kana Abe")
  })

  it("renders the date", () => {
    const date = renderer.create(<Date date="2017-05-19T13:09:18.567Z" layout="split" />)
    expect(date).toMatchSnapshot()
    const shallowDate = shallow(<Date date="2017-05-19T13:09:18.567Z" layout="split" />)
    expect(shallowDate.html()).toContain("May 19, 2017 9:09 am")
  })
})
