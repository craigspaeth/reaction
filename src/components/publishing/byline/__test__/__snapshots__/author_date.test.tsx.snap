// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author/Date renders a single author 1`] = `
.c0 {
  font-family: Unica77LLWebMedium ,      'Arial',      'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 16px;
  line-height: 1.1em;
  margin: 10px 30px 0 0;
}

.c0.date {
  white-space: nowrap;
}

.c0:before {
  content: "";
  display: inline-block;
  min-width: 10px;
  min-height: 10px;
  border-radius: 50%;
  margin: 6px 10px 0 0;
  background-color: #000;
}

@media (max-width: 720px) {
  .c0 {
    font-family: Unica77LLWebMedium ,      'Arial',      'serif';
    -webkit-font-smoothing: antialiased;
    font-size: 14px;
    line-height: 1.4em;
    font-family: Unica77LLWebMedium ,      'Arial',      'serif';
    -webkit-font-smoothing: antialiased;
    font-size: 14px;
    line-height: 1.4em;
    margin: 10px 20px 0 0;
  }
}

@media (max-width: 720px) {
  .c0:before {
    min-width: 8px;
    min-height: 8px;
  }
}

<div
  className="author c0"
>
  By 
  Molly Gottschalk
</div>
`;

exports[`Author/Date renders multiple authors 1`] = `
.c0 {
  font-family: Unica77LLWebMedium ,      'Arial',      'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 16px;
  line-height: 1.1em;
  margin: 10px 30px 0 0;
}

.c0.date {
  white-space: nowrap;
}

.c0:before {
  content: "";
  display: inline-block;
  min-width: 10px;
  min-height: 10px;
  border-radius: 50%;
  margin: 6px 10px 0 0;
  background-color: #000;
}

@media (max-width: 720px) {
  .c0 {
    font-family: Unica77LLWebMedium ,      'Arial',      'serif';
    -webkit-font-smoothing: antialiased;
    font-size: 14px;
    line-height: 1.4em;
    font-family: Unica77LLWebMedium ,      'Arial',      'serif';
    -webkit-font-smoothing: antialiased;
    font-size: 14px;
    line-height: 1.4em;
    margin: 10px 20px 0 0;
  }
}

@media (max-width: 720px) {
  .c0:before {
    min-width: 8px;
    min-height: 8px;
  }
}

<div
  className="author c0"
>
  By 
  Molly Gottschalk and Kana Abe
</div>
`;

exports[`Author/Date renders the date 1`] = `
.c0 {
  font-family: Unica77LLWebMedium ,      'Arial',      'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 16px;
  line-height: 1.1em;
  margin: 10px 30px 0 0;
}

.c0.date {
  white-space: nowrap;
}

@media (max-width: 720px) {
  .c0 {
    font-family: Unica77LLWebMedium ,      'Arial',      'serif';
    -webkit-font-smoothing: antialiased;
    font-size: 14px;
    line-height: 1.4em;
    font-family: Unica77LLWebMedium ,      'Arial',      'serif';
    -webkit-font-smoothing: antialiased;
    font-size: 14px;
    line-height: 1.4em;
    margin: 10px 20px 0 0;
  }
}

<div
  className="date c0"
>
  May 19, 2017 9:09 am
</div>
`;
